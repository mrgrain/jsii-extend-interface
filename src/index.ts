export * from './jsii-struct-builder';
export * from './jsii-struct-file';
export * from './typescript-renderer';
