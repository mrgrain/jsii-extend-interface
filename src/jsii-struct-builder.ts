import { InterfaceType, Property } from '@jsii/spec';

/**
 * Build a jsii struct
 */
export class JsiiStructBuilder {
  private targetSpec: InterfaceType;
  private properties: Map<string, Property>;

  public constructor(base: InterfaceType) {
    this.targetSpec = base;
    this.properties = new Map(base.properties?.map((p) => [p.name, p]));
  }

  /**
   * Omit this prop
   */
  public omit(prop: string): JsiiStructBuilder {
    this.properties.delete(prop);

    return this;
  }

  /**
   * Add a prop
   */
  public add(prop: Property): JsiiStructBuilder {
    this.properties.set(prop.name, prop);

    return this;
  }

  /**
   * Update an existing prop
   */
  public update(name: string, update: Partial<Property>): JsiiStructBuilder {
    const old = this.properties.get(name);

    if (!old) {
      throw `Unable top update property '${name}' in '${this.targetSpec.fqn}: Property does not exists, please use add'`;
    }

    const updated = {
      ...old,
      ...update,
      docs: {
        ...old?.docs,
        ...update?.docs,
        custom: {
          ...old?.docs?.custom,
          ...update?.docs?.custom,
        },
      },
    };

    return this.add(updated);
  }

  /**
   * Get the current state of the builder
   */
  public get(): InterfaceType {
    this.targetSpec.properties = Array.from(this.properties.values());

    return this.targetSpec;
  }
}
