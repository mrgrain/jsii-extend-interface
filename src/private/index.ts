export * from './assembly';
export * from './utils';
