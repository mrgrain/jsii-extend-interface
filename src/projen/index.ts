export * from './projen-struct';
export * from './ts-interface';
