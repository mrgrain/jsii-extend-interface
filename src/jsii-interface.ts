import { dirname, join, posix } from 'path';
import { InterfaceType, Property, TypeKind } from '@jsii/spec';
import { Component, typescript } from 'projen';
import { findInterface } from './assembly';
import { InterfaceFile } from './interface-file';

export interface JsiiInterfaceOptions {
  /**
   * The name of the interface
   */
  readonly name: string;
  /**
   * Doc string for the interface
   * @default - Interface name
   */
  readonly description?: string;
  /**
   * The fqn of the interface
   *
   * Used to auto-add imports.
   * All referenced types are loaded based on the fqn hierarchy.
   *
   * See `importLocations` for customization options.
   *
   * @default `${project.name}.${options.name}`
   */
  readonly fqn?: string;
  /**
   * Output file path
   * @default - inside `srcdir` or `outdir` with the fqn turned into a path
   */
  readonly filePath?: string;
  /**
   * The module locations assemblies should be imported from
   *
   * All types are imported from the top-level of the module.
   * This map can be used to overwrite the default import locations.
   *
   * For local imports, the import traverse up a number of levels equivalent to the number of fqn parts of the rendered type.
   * E.g. if the the rendered type is `pkg.interface`, it is assumed to be at the top-level and imports from the same assembly would be from `./`.
   * If the rendered type is `pkg.nested.sub.interface` a local import will be from `../../`.
   *
   * @default - uses the assembly name for external packages
   * local imports traverse up a number of levels equivalent to the number of fqn parts
   */
  importLocations?: Record<string, string>;
  /**
   * The properties of this interface
   *
   * When extending, these will replace existing properties.
   *
   * @default []
   */
  readonly props?: {
    [name: string]: Omit<Property, 'name'>;
  };
  /**
   * Extends this jsii interface
   */
  readonly extends?: string;
  /**
   * Omit these properties from the parent interface
   */
  readonly omitProps?: string[];
  /**
   * Update these properties from the parent interface
   *
   * Property definitions are merged non-recursively.
   * I.e. when updating `type` a complete new type must be provided
   */
  readonly updateProps?: {
    [name: string]: Partial<Property>;
  };
}

/**
 * A component generating a jsii-compatible interface
 */
export class JsiiInterface extends Component {
  public constructor(
    project: typescript.TypeScriptProject,
    options: JsiiInterfaceOptions
  ) {
    super(project);

    const properties = Object.entries(options.props || {}).map(([name, p]) => ({
      name,
      ...p,
    }));

    const fqn = options.fqn ?? `${project.name}.${options.name}`;
    const assembly = fqn.split('.').at(0) ?? project.name;
    const targetSpec: InterfaceType = {
      kind: TypeKind.Interface,
      assembly,
      fqn,
      name: options.name,
      docs: {
        summary: options.description ?? options.name,
      },
      properties,
    };

    if (options.extends) {
      const sourceSpec = findInterface(options.extends, true);
      const omit = [
        ...(options.omitProps || []),
        ...properties.map((p) => p.name),
      ];
      targetSpec.properties!.push(
        ...(sourceSpec.properties || [])
          .map((p) => {
            if (options.updateProps?.[p.name]) {
              return {
                ...p,
                ...options.updateProps?.[p.name],
                docs: {
                  ...p.docs,
                  ...options.updateProps?.[p.name].docs,
                  custom: {
                    ...p.docs?.custom,
                    ...options.updateProps?.[p.name].docs?.custom,
                  },
                },
              };
            }
            return p;
          })
          .filter((p) => !omit.includes(p.name))
      );
    }

    targetSpec.properties = targetSpec.properties?.sort((a, b) =>
      a.name.localeCompare(b.name)
    );

    const baseDir = project.srcdir ?? project.outdir;
    const outputFile = options.filePath ?? join(baseDir, fqnToPath(fqn));
    new InterfaceFile(project, outputFile, targetSpec, {
      importLocations: {
        [assembly]: relativeImport(outputFile, baseDir),
        ...options.importLocations,
      },
    });
  }
}
function relativeImport(from: string, target: string): string {
  const diff = posix.relative(dirname(from), target);
  if (!diff) {
    return '.' + posix.sep;
  }
  return diff + posix.sep;
}

function fqnToPath(fqn: string): string {
  return join(...fqn.split('.').slice(1)) + '.ts';
}
