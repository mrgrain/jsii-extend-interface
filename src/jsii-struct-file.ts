import { InterfaceType } from '@jsii/spec';
import { Project, SourceCodeOptions, TextFile } from 'projen';
import {
  TypeScriptRenderer,
  TypeScriptRendererOptions,
} from './typescript-renderer';

/**
 * Options for `JsiiStructFile`.
 */
export interface JsiiStructFileOptions
  extends TypeScriptRendererOptions,
    SourceCodeOptions {}

/**
 * A Jsii Struct rendered as TypeScript interface
 */
export class JsiiStructFile extends TextFile {
  public constructor(
    project: Project,
    filePath: string,
    spec: InterfaceType,
    options: JsiiStructFileOptions = {}
  ) {
    super(project, filePath, options);

    const renderer = new TypeScriptRenderer(options);
    this.addLine(`// ${this.marker}`);
    this.addLine(renderer.renderStruct(spec));
  }
}
