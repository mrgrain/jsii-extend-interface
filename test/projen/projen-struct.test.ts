import { join } from 'path';
import { PrimitiveType, Stability } from '@jsii/spec';
import {
  TypeScriptProject,
  TypeScriptProjectOptions,
} from 'projen/lib/typescript';
import { synthSnapshot } from 'projen/lib/util/synth';
import { Struct, ProjenStruct } from '../../src';

test('can mixin a struct', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const struct = new ProjenStruct(project, {
    name: 'MyInterface',
  });
  struct.mixin(Struct.fromFqn('projen.typescript.ProjenrcOptions'));

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toMatchSnapshot();
  expect(renderedFile).toContain('projenCodeDir');
  expect(renderedFile).toContain('filename');
});

test('can mixin another projen struct', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const base = new ProjenStruct(project, {
    name: 'MyBaseInterface',
  });
  base.mixin(Struct.fromFqn('projen.typescript.ProjenrcOptions'));

  const struct = new ProjenStruct(project, {
    name: 'MyInterface',
  });
  struct.mixin(base);

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toMatchSnapshot();
  expect(renderedFile).toContain('projenCodeDir');
  expect(renderedFile).toContain('filename');
});

test('can omit props', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const struct = new ProjenStruct(project, {
    name: 'MyInterface',
  });
  struct
    .mixin(Struct.fromFqn('projen.typescript.ProjenrcOptions'))
    .omit('projenCodeDir', 'filename');

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).not.toContain('projenCodeDir');
  expect(renderedFile).not.toContain('filename');
});

test('can keep only some props', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const struct = new ProjenStruct(project, {
    name: 'MyInterface',
  });
  struct
    .mixin(Struct.fromFqn('projen.typescript.ProjenrcOptions'))
    .only('projenCodeDir');

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toContain('projenCodeDir');
  expect(renderedFile).not.toContain('filename');
});

test('can ignore deprecated props', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const struct = new ProjenStruct(project, {
    name: 'MyInterface',
  });
  struct.add(
    {
      name: 'currentProp',
      type: { primitive: PrimitiveType.Boolean },
    },
    {
      name: 'deprecatedProps',
      type: { primitive: PrimitiveType.Boolean },
      docs: { deprecated: 'use `currentProp`' },
    }
  );
  struct.withoutDeprecated();

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).not.toContain('deprecatedProps');
  expect(renderedFile).toContain('currentProp');
});

test('can make all props optional', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const struct = new ProjenStruct(project, { name: 'MyInterface' });
  struct.add(
    {
      name: 'optionalProp',
      type: { primitive: PrimitiveType.Boolean },
      optional: true,
    },
    {
      name: 'requiredProp',
      type: { primitive: PrimitiveType.Boolean },
      optional: false,
    }
  );
  struct.allOptional();

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toContain('optionalProp?: boolean');
  expect(renderedFile).toContain('requiredProp?: boolean');
});

test('can overwrite props', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const struct = new ProjenStruct(project, {
    name: 'MyInterface',
  });
  struct.mixin(Struct.fromFqn('projen.typescript.ProjenrcOptions')).add({
    name: 'projenCodeDir',
    type: {
      primitive: PrimitiveType.Number,
    },
  });

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toContain('projenCodeDir: number');
});

test('can update props', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const struct = new ProjenStruct(project, {
    name: 'MyInterface',
  });
  struct
    .mixin(Struct.fromFqn('projen.typescript.ProjenrcOptions'))
    .update('projenCodeDir', {
      docs: {
        summary: 'New summary',
        stability: Stability.Stable,
        custom: {
          pjnew: '"newVal"',
        },
      },
      optional: false,
    });

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toContain('New summary');
  expect(renderedFile).toContain('@stability stable');
  expect(renderedFile).toContain('@default "projenrc"');
  expect(renderedFile).toContain('@pjnew "newVal"');
  expect(renderedFile).toMatchSnapshot();
});

test('can updateAll props', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const struct = new ProjenStruct(project, {
    name: 'MyInterface',
  });
  struct.mixin(Struct.fromFqn('projen.typescript.ProjenrcOptions')).updateAll({
    docs: {
      stability: Stability.Stable,
    },
  });

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toContain('@stability stable');
  expect(renderedFile).toMatchSnapshot();
});

test('can import type from the same package at the top level', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const struct = new ProjenStruct(project, {
    name: 'MyInterface',
    fqn: 'test.MyInterface',
  });
  struct.add(
    {
      name: 'localProp',
      type: {
        fqn: 'test.OtherInterface',
      },
    },
    {
      name: 'localNestedProp',
      type: {
        fqn: 'test.more.OtherInterface',
      },
    }
  );

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toContain("import { more, OtherInterface } from './';");
});

test('can import type from the same package when nested', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const struct = new ProjenStruct(project, {
    name: 'MyInterface',
    fqn: 'test.a.b.c.MyInterface',
  });
  struct.add(
    {
      name: 'localProp',
      type: {
        fqn: 'test.OtherInterface',
      },
    },
    {
      name: 'localNestedProp',
      type: {
        fqn: 'test.more.OtherInterface',
      },
    }
  );

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/a/b/c/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toContain(
    "import { more, OtherInterface } from '../../../';"
  );
});

test("can import type from the same package when in a location that's not matching fqn levels", () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const struct = new ProjenStruct(project, {
    name: 'MyInterface',
    fqn: 'test.one.two.three.MyInterface',
    filePath: join(project.srcdir, 'sub', 'MyInterface.ts'),
  });
  struct.add(
    {
      name: 'localProp',
      type: {
        fqn: 'test.OtherInterface',
      },
    },
    {
      name: 'localNestedProp',
      type: {
        fqn: 'test.more.OtherInterface',
      },
    }
  );

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/sub/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toContain("import { more, OtherInterface } from '../';");
});

test('can import type from an other package', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const struct = new ProjenStruct(project, {
    name: 'MyInterface',
    fqn: 'test.MyInterface',
  });

  struct.mixin(Struct.fromFqn('projen.typescript.ProjenrcOptions')).add({
    name: 'testProp',
    type: {
      fqn: 'projen.typescript.TypeScriptProjectOptions',
    },
  });

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toContain("import { typescript } from 'projen';");
});

test('can override import locations', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const struct = new ProjenStruct(project, {
    name: 'MyInterface',
    fqn: 'test.MyInterface',
    importLocations: {
      foo: 'bar',
      projen: 'banana',
    },
  });

  struct.mixin(Struct.fromFqn('projen.typescript.ProjenrcOptions')).add(
    {
      name: 'localProp',
      type: {
        fqn: 'foo.OtherInterface',
      },
    },
    {
      name: 'localNestedProp',
      type: {
        fqn: 'foo.more.OtherInterface',
      },
    },
    {
      name: 'externalProp',
      type: {
        fqn: 'projen.typescript.TypeScriptProjectOptions',
      },
    }
  );

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toContain("import { more, OtherInterface } from 'bar';");
  expect(renderedFile).toContain("import { typescript } from 'banana';");
});

test('can use struct as type in add', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const base = new ProjenStruct(project, {
    name: 'MyBaseInterface',
  });

  const nestedStruct = new ProjenStruct(project, {
    name: 'MyInterface',
  });
  nestedStruct
    .mixin(Struct.fromFqn('projen.typescript.ProjenrcOptions'))
    .update('filename', { docs: { default: '.projenRC.ts' } });

  base.add({
    name: 'projenrcTsOptions',
    type: nestedStruct,
    docs: {
      summary: 'My new Summary',
    },
  });

  // PREPARE
  const synthSnapshotOutput = synthSnapshot(project);
  const renderedNestedFile = synthSnapshotOutput['src/MyInterface.ts'];
  const renderedBaseFile = synthSnapshotOutput['src/MyBaseInterface.ts'];

  // ASSERT
  expect(renderedNestedFile).toMatchSnapshot();
  expect(renderedNestedFile).toContain('@default .projenRC.ts');
  expect(renderedBaseFile).toMatchSnapshot();
  expect(renderedBaseFile).toContain("import { MyInterface } from './'");
  expect(renderedBaseFile).toContain(
    'readonly projenrcTsOptions: MyInterface;'
  );
});

test('can use struct as type in update', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const base = new ProjenStruct(project, {
    name: 'MyBaseInterface',
  });

  const nestedStruct = new ProjenStruct(project, {
    name: 'MyInterface',
  });
  nestedStruct
    .mixin(Struct.fromFqn('projen.typescript.ProjenrcOptions'))
    .update('filename', { docs: { default: '.projenRC.ts' } });

  base
    .mixin(Struct.fromFqn('projen.typescript.TypeScriptProjectOptions'))
    .update('projenrcTsOptions', {
      type: nestedStruct,
      docs: {
        summary: 'My new Summary',
      },
    });

  // PREPARE
  const synthSnapshotOutput = synthSnapshot(project);
  const renderedNestedFile = synthSnapshotOutput['src/MyInterface.ts'];
  const renderedBaseFile = synthSnapshotOutput['src/MyBaseInterface.ts'];

  // ASSERT
  expect(renderedNestedFile).toContain('@default .projenRC.ts');
  expect(renderedBaseFile).toContain("import { MyInterface } from './'");
  expect(renderedBaseFile).toContain(
    'readonly projenrcTsOptions?: MyInterface'
  );
});

class TestProject extends TypeScriptProject {
  public constructor(options: Partial<TypeScriptProjectOptions> = {}) {
    super({
      name: 'test',
      defaultReleaseBranch: 'main',
      ...options,
    });
  }
}

test('can create a struct from empty', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const base = new ProjenStruct(project, {
    name: 'MyInterface',
  });
  const addedStruct = Struct.empty();
  addedStruct.add({
    name: 'emptyProp',
    type: { primitive: PrimitiveType.Boolean },
  });
  base.mixin(addedStruct);

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toContain('emptyProp');
});

test('can use an empty struct as type with name', () => {
  // ARRANGE
  const project = new TestProject();

  // ACT
  const nestedStruct = Struct.empty('pkg.sub.MyOtherInterface');
  nestedStruct.add({
    name: 'emptyProp',
    type: { primitive: PrimitiveType.Boolean },
  });

  const base = new ProjenStruct(project, {
    name: 'MyInterface',
  });
  base.add({
    name: 'nestedProp',
    type: nestedStruct,
  });

  // PREPARE
  const renderedFile = synthSnapshot(project)['src/MyInterface.ts'];

  // ASSERT
  expect(renderedFile).toContain("import { sub } from 'pkg';");
  expect(renderedFile).toContain('readonly nestedProp: sub.MyOtherInterface');
});
