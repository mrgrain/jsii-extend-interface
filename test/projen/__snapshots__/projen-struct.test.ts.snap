// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can mixin a struct 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

/**
 * MyInterface
 */
export interface MyInterface {
  /**
   * Whether to use \`SWC\` for ts-node.
   * @default false
   * @stability experimental
   */
  readonly swc?: boolean;
  /**
   * A directory tree that may contain *.ts files that can be referenced from your projenrc typescript file.
   * @default "projenrc"
   * @stability experimental
   */
  readonly projenCodeDir?: string;
  /**
   * The name of the projenrc file.
   * @default ".projenrc.ts"
   * @stability experimental
   */
  readonly filename?: string;
}
"
`;

exports[`can mixin another projen struct 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

/**
 * MyInterface
 */
export interface MyInterface {
  /**
   * The name of the projenrc file.
   * @default ".projenrc.ts"
   * @stability experimental
   */
  readonly filename?: string;
  /**
   * A directory tree that may contain *.ts files that can be referenced from your projenrc typescript file.
   * @default "projenrc"
   * @stability experimental
   */
  readonly projenCodeDir?: string;
  /**
   * Whether to use \`SWC\` for ts-node.
   * @default false
   * @stability experimental
   */
  readonly swc?: boolean;
}
"
`;

exports[`can update props 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

/**
 * MyInterface
 */
export interface MyInterface {
  /**
   * Whether to use \`SWC\` for ts-node.
   * @default false
   * @stability experimental
   */
  readonly swc?: boolean;
  /**
   * New summary
   * @stability stable
   * @pjnew "newVal"
   */
  readonly projenCodeDir: string;
  /**
   * The name of the projenrc file.
   * @default ".projenrc.ts"
   * @stability experimental
   */
  readonly filename?: string;
}
"
`;

exports[`can updateAll props 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

/**
 * MyInterface
 */
export interface MyInterface {
  /**
   * Whether to use \`SWC\` for ts-node.
   * @default false
   * @stability stable
   */
  readonly swc?: boolean;
  /**
   * A directory tree that may contain *.ts files that can be referenced from your projenrc typescript file.
   * @default "projenrc"
   * @stability stable
   */
  readonly projenCodeDir?: string;
  /**
   * The name of the projenrc file.
   * @default ".projenrc.ts"
   * @stability stable
   */
  readonly filename?: string;
}
"
`;

exports[`can use struct as type in add 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

/**
 * MyInterface
 */
export interface MyInterface {
  /**
   * Whether to use \`SWC\` for ts-node.
   * @default false
   * @stability experimental
   */
  readonly swc?: boolean;
  /**
   * A directory tree that may contain *.ts files that can be referenced from your projenrc typescript file.
   * @default "projenrc"
   * @stability experimental
   */
  readonly projenCodeDir?: string;
  /**
   * The name of the projenrc file.
   * @default .projenRC.ts
   * @stability experimental
   */
  readonly filename?: string;
}
"
`;

exports[`can use struct as type in add 2`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
import { MyInterface } from './';

/**
 * MyBaseInterface
 */
export interface MyBaseInterface {
  /**
   * My new Summary
   */
  readonly projenrcTsOptions: MyInterface;
}
"
`;
